### PR TITLE
Eka 25: SMS Notification

### PR DESCRIPTION
**Pull request by Nexengineer**

### CODE REVIEW

File: src/main/java/in/projecteka/consentmanager/consent/model/NotificationAction.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package in.projecteka.consentmanager.consent.model;
+
+import com.fasterxml.jackson.annotation.JsonValue;
+
+public enum NotificationAction {
+    CONSENT_REQUEST("ConsentRequest");
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/model/NotificationAction.java]:
```suggestion
    CONSENT_REQUEST_CREATED("ConsentRequestCreated");
```

Comment (Nexengineer) [reply src/main/java/in/projecteka/consentmanager/consent/model/NotificationAction.java]: Done. Codacy issue is unused listner

File: src/main/java/in/projecteka/consentmanager/clients/ConsentNotificationClient.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    private final WebClient.Builder webClientBuilder;
+
+    public Mono<Void> sendToNotificationService(NotificationMessage notificationMessage) {
+        return webClientBuilder.build()
+                .post()
+                .uri(otpServiceProperties.getUrl() + "/notification")
+                .header("Content-Type", MediaType.APPLICATION_JSON_VALUE)
+                .bodyValue(notificationMessage)
+                .retrieve()
+                .onStatus(HttpStatus::is5xxServerError, clientResponse -> Mono.error(unknownErrorOccurred()))
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/clients/ConsentNotificationClient.java]: use isError now, as we have to return `unknownErrorOccured'

File: src/main/java/in/projecteka/consentmanager/consent/ConsentConfiguration.java [hunk 3]
```diff
@@ -37,21 +39,29 @@ public PostConsentApproval postConsentApproval(AmqpTemplate amqpTemplate, Destin
         return new PostConsentApproval(amqpTemplate, destinationsConfig);
     }
 
+    @Bean
+    public PostConsentRequestNotification postConsentRequestNotification(AmqpTemplate amqpTemplate,
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentConfiguration.java]: Why do we need Notification suffix?

File: src/main/java/in/projecteka/consentmanager/consent/ConsentConfiguration.java [hunk 4]
```diff
@@ -84,6 +94,26 @@ public ConsentArtefactBroadcastListener hiuNotificationListener(MessageListenerC
...
+    @Bean
+    public ConsentRequestNotificationListener consentRequestNotificationListener(
+                                                                    MessageListenerContainerFactory messageListenerContainerFactory,
+                                                                    DestinationsConfig destinationsConfig,
+                                                                    Jackson2JsonMessageConverter jackson2JsonMessageConverter,
+                                                                    WebClient.Builder builder,
+                                                                    OtpServiceProperties otpServiceProperties,
+                                                                    UserServiceProperties userServiceProperties,
+                                                                    ConsentServiceProperties consentServiceProperties
+                                                                    ) {
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentConfiguration.java]:
```suggestion
          ) {
```
```suggestion
    @Bean
    public ConsentRequestNotificationListener consentRequestNotificationListener(
                                                                    MessageListenerContainerFactory messageListenerContainerFactory,
                                                                    DestinationsConfig destinationsConfig,
                                                                    Jackson2JsonMessageConverter jackson2JsonMessageConverter,
                                                                    WebClient.Builder builder,
                                                                    OtpServiceProperties otpServiceProperties,
                                                                    UserServiceProperties userServiceProperties,
                                                                    ConsentServiceProperties consentServiceProperties) {
```

File: src/main/java/in/projecteka/consentmanager/consent/ConsentConfiguration.java [hunk 5]
```diff
@@ -84,6 +94,26 @@ public ConsentArtefactBroadcastListener hiuNotificationListener(MessageListenerC
...
+        return new ConsentRequestNotificationListener(
+                messageListenerContainerFactory,
+                destinationsConfig,
+                jackson2JsonMessageConverter,
+                new ConsentNotificationClient(otpServiceProperties, builder),
+                new UserServiceClient(builder, userServiceProperties),
+                consentServiceProperties
+        );
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentConfiguration.java]:
```suggestion
        return new ConsentRequestNotificationListener(
                messageListenerContainerFactory,
                destinationsConfig,
                jackson2JsonMessageConverter,
                new ConsentNotificationClient(otpServiceProperties, builder),
                new UserServiceClient(builder, userServiceProperties),
                consentServiceProperties);
```

File: src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java [hunk 6]
```diff
@@ -57,9 +59,18 @@ private static boolean isNotSameRequester(ConsentArtefact consentDetail, String
                 .flatMap(context -> validatePatient(requestedDetail.getPatient().getId(), context.get("Authorization")))
                 .then(validateHIPAndHIU(requestedDetail))
                 .then(saveRequest(requestedDetail, requestId))
+                .then(sendConsentRequestNotification(requestId, requestedDetail))
                 .thenReturn(requestId);
     }
 
+    private Mono<Void> sendConsentRequestNotification(String requestId, RequestedDetail requestedDetail) {
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java]:
```suggestion
    private Mono<Void> sendConsentRequestNotification(String requestId, RequestedDetail requestedDetail) {
```

No need of this method I believe.

File: src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java [hunk 7]
```diff
@@ -57,9 +59,18 @@ private static boolean isNotSameRequester(ConsentArtefact consentDetail, String
                 .flatMap(context -> validatePatient(requestedDetail.getPatient().getId(), context.get("Authorization")))
                 .then(validateHIPAndHIU(requestedDetail))
                 .then(saveRequest(requestedDetail, requestId))
+                .then(sendConsentRequestNotification(requestId, requestedDetail))
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentManager.java]: ConsentManager does not need to know what is our post processor do with consentRequest.

File: src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+            ConsentRequest consentRequest =
+                    (ConsentRequest) converter.fromMessage(message);
+            logger.info(String.format(
+                    "Received message for Request id : %s", consentRequest.getRequestId()));
+            createNotificationMessage(consentRequest)
+                    .flatMap(this::callNotificationService)
+                    .block();
+        };
+        mlc.setupMessageListener(messageListener);
+        mlc.start();
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java]:
```suggestion
        MessageListenerContainer mlc = messageListenerContainerFactory
                .createMessageListenerContainer(destinationInfo.getRoutingKey());

        MessageListener messageListener = message -> {
            var consentRequest = (ConsentRequest) converter.fromMessage(message);
            logger.info(String.format("Received message for Request id : %s", consentRequest.getRequestId()));
            createNotificationMessage(consentRequest)
                    .flatMap(this::callNotificationService)
                    .block();
        };
        mlc.setupMessageListener(messageListener);
        mlc.start();
```

File: src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    public Mono<Void> callNotificationService(NotificationMessage notificationMessage) {
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java]:
```suggestion
    public Mono<Void> NotifyUserWith(NotificationMessage notificationMessage) {
```

File: src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        return consentNotificationClient.sendToNotificationService(notificationMessage);
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java]:
```suggestion
        return consentNotificationClient.send(notificationMessage);
```

File: src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    private Mono<NotificationMessage> createNotificationMessage(ConsentRequest consentRequest) {
+        return userServiceClient.userOf(consentRequest.getRequestedDetail().getPatient().getId())
+                .flatMap(user -> Mono.just(NotificationMessage.builder()
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java]:
```suggestion
                .map(user -> NotificationMessage.builder()
```

File: src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+                                .requester(consentRequest.getRequestedDetail().getRequester().getName())
+                                .consentRequestId(consentRequest.getRequestId())
+                                .hiTypes(Arrays.stream(consentRequest.getRequestedDetail().getHiTypes())
+                                        .map(HIType::getValue)
+                                        .collect(Collectors.joining(",")))
+                                .deepLinkUrl(String.format("%s/consent/%s",
+                                        consentServiceProperties.getConsentServiceUrl(),
+                                        consentRequest.getRequestId()))
+                                .build())
+                        .build()));
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java]:
```suggestion
                        .build());
```

File: src/main/java/in/projecteka/consentmanager/consent/ConsentServiceProperties.java [hunk 13]
```diff
@@ -17,4 +17,5 @@
 public class ConsentServiceProperties {
     private int maxPageSize;
     private int defaultPageSize;
+    private String consentServiceUrl;
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/ConsentServiceProperties.java]:
```suggestion
    private String consentServiceUrl;
```
Can this be just an URL?

File: src/main/java/in/projecteka/consentmanager/consent/PostConsentRequestNotification.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+                destinationsConfig.getQueues().get(CONSENT_REQUEST_QUEUE);
+        if (destinationInfo == null) {
+            logger.info(CONSENT_REQUEST_QUEUE + " not found");
+            throw queueNotFound();
+        }
+        return Mono.create(monoSink -> {
+            amqpTemplate.convertAndSend(
+                    destinationInfo.getExchange(),
+                    destinationInfo.getRoutingKey(),
+                    consentRequest);
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/PostConsentRequestNotification.java]:
```suggestion
            amqpTemplate.convertAndSend(destinationInfo.getExchange(), destinationInfo.getRoutingKey(), consentRequest);
```

Comment (ganesanarun) [reply src/main/java/in/projecteka/consentmanager/consent/PostConsentRequestNotification.java]: If it's within 120 characters limit or whatever default you have.

File: src/main/java/in/projecteka/consentmanager/consent/model/ConsentRequest.java [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@AllArgsConstructor
+@Data
+@Builder
+@NoArgsConstructor
+public class ConsentRequest {
+    private String requestId;
+    private RequestedDetail requestedDetail;
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/model/ConsentRequest.java]:
```suggestion
    private String id;
    private RequestedDetail detail;
```

File: src/main/java/in/projecteka/consentmanager/consent/model/NotificationMessage.java [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
+public class NotificationMessage {
+    private String id;
+    private Communication communication;
+    private NotificationContent notificationContent;
+    private NotificationAction notificationAction;
+}
```
Comment (ganesanarun) [inline src/main/java/in/projecteka/consentmanager/consent/model/NotificationMessage.java]:
```suggestion
@Data
@AllArgsConstructor
@NoArgsConstructor
@Builder
public class Notification {
    private String id;
    private Communication communication;
    private Content content;
    private Action action;
}
```

File: src/main/java/in/projecteka/consentmanager/clients/ConsentNotificationClient.java (hunk 2 above)
Comment (Nexengineer) [reply src/main/java/in/projecteka/consentmanager/clients/ConsentNotificationClient.java]: Done

File: src/main/java/in/projecteka/consentmanager/consent/ConsentConfiguration.java (hunk 3 above)
Comment (Nexengineer) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentConfiguration.java]: Changed it to PostConsentRequest

File: src/main/java/in/projecteka/consentmanager/consent/ConsentConfiguration.java (hunk 4 above)
Comment (Nexengineer) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentConfiguration.java]: Done

File: src/main/java/in/projecteka/consentmanager/consent/ConsentConfiguration.java (hunk 5 above)
Comment (Nexengineer) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentConfiguration.java]: Done

File: src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java (hunk 8 above)
Comment (Nexengineer) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java]: Done

File: src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java (hunk 9 above)
Comment (Nexengineer) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java]: Done

File: src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java (hunk 10 above)
Comment (Nexengineer) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java]: Done

File: src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java (hunk 11 above)
Comment (Nexengineer) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java]: Done

File: src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java (hunk 12 above)
Comment (Nexengineer) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentRequestNotificationListener.java]: Done

File: src/main/java/in/projecteka/consentmanager/consent/ConsentServiceProperties.java (hunk 13 above)
Comment (Nexengineer) [reply src/main/java/in/projecteka/consentmanager/consent/ConsentServiceProperties.java]: Done

File: src/main/java/in/projecteka/consentmanager/consent/PostConsentRequestNotification.java (hunk 14 above)
Comment (Nexengineer) [reply src/main/java/in/projecteka/consentmanager/consent/PostConsentRequestNotification.java]: Done

File: src/main/java/in/projecteka/consentmanager/consent/model/ConsentRequest.java (hunk 15 above)
Comment (Nexengineer) [reply src/main/java/in/projecteka/consentmanager/consent/model/ConsentRequest.java]: done

File: src/main/java/in/projecteka/consentmanager/consent/model/NotificationMessage.java (hunk 16 above)
Comment (Nexengineer) [reply src/main/java/in/projecteka/consentmanager/consent/model/NotificationMessage.java]: Done

**Pull request merged**